### ISSUE TITLE
"JSONValue is not an object" crash in `UploadSession.restore()

### ISSUE THREAD

**Issue by eddy-geek** [abraunegg/onedrive#663]
### Bug Report Details ###
**Describe the bug**

Immediate "JSONValue is not an object" crash in `src/upload.d:63 bool upload.UploadSession.restore()`

I'm using master so it should not be #550, not the exact same stack as others like #464 ...

**Application and Operating System Details:**
*   OS: Output of `uname -a` & provide your OS & version (CentOS 6.x, Ubuntu 18.x etc)
Fedora 30
```
Linux ncelrnd0216 5.2.11-200.fc30.x86_64 #1 SMP Thu Aug 29 12:43:20 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```
*   Are you using a headless system (no gui) or with a gui installed?
KDE gui
*   OneDrive Account Type
office365, shared university account (so, business)
*   DMD or LDC compiler version `dmd --version` or `ldmd2 --version`
DMD64 D Compiler v2.087.0

*   Application configuration: Output of `onedrive --display-config`
```
onedrive version                    = v2.3.9-11-g2abc05a
Config path                         = /home/ed/.config/onedrive
Config file found in config path    = true
Config option 'check_nosync'        = false
Config option 'sync_dir'            = /ything/EveOneDrive
Config option 'skip_dir'            = 
Config option 'skip_file'           = ~|.~*|*.tmp
Config option 'skip_dotfiles'       = false
Config option 'skip_symlinks'       = false
Config option 'monitor_interval'    = 45
Config option 'min_notify_changes'  = 1
Config option 'log_dir'             = /var/log/onedrive/
Config option 'sync_root_files'     = false
Selective sync configured           = false
```
*   Curl Version: Output of `curl --version`
```
curl 7.65.3 (x86_64-redhat-linux-gnu) libcurl/7.65.3 OpenSSL/1.1.1c-fips zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.20.2 (+libidn2/2.0.5) libssh/0.9.0/openssl/zlib nghttp2/1.39.2
```
**Complete Verbose Log Output**
A clear and full log of the problem when running the application in the following manner (ie, not in monitor mode):
```bash
onedrive --synchronize --verbose
Using Config Dir: /home/ed/.config/onedrive
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /ything/EveOneDrive
Application version: v2.3.9-11-g2abc05a
Account Type: business
Default Drive ID: b!qS4f0OIpakGljF0Q3eFGdU849ksdmgtGulX34TgjurUU_4eLzWBeR5k-SFaEiw8t
Default Root ID: 01FGSPRTN6Y2GOVW7725BZO354PWSELRRZ
Remaining Free Space: 987838457485
Fetching details for OneDrive Root
OneDrive Root exists in the database
Trying to restore the upload session ...
std.json.JSONException@std/json.d(304): JSONValue is not an object
----------------
/home/ed/dlang/dmd-2.087.0/linux/bin64/../../src/phobos/std/exception.d:515 pure @safe void std.exception.bailOut!(std.json.JSONException).bailOut(immutable(char)[], ulong, scope const(char)[]) [0x6086e3]
/home/ed/dlang/dmd-2.087.0/linux/bin64/../../src/phobos/std/exception.d:436 pure @safe bool std.exception.enforce!(std.json.JSONException).enforce!(bool).enforce(bool, lazy const(char)[], immutable(char)[], ulong) [0x60865b]
??:? inout pure @property @trusted inout(std.json.JSONValue[immutable(char)[]]) std.json.JSONValue.objectNoRef() [0x678571]
/home/ed/dlang/dmd-2.087.0/linux/bin64/../../src/phobos/std/json.d:643 const pure @safe const(std.json.JSONValue)* std.json.JSONValue.opBinaryRight!("in").opBinaryRight(immutable(char)[]) [0x627bff]
src/upload.d:63 bool upload.UploadSession.restore() [0x64292e]
src/sync.d:363 void sync.SyncEngine.init() [0x634adb]
src/main.d:755 bool main.initSyncEngine(sync.SyncEngine) [0x6228fc]
src/main.d:522 _Dmain [0x621647]
```

### Bug Report Checklist ###
*   [X] Detailed description
*   [X] Reproduction steps (if applicable)
*   [X] Verbose Log Output

**Comment by abraunegg** [comment]
@eddy-geek 
Thanks .. will look into this shortly

**Comment by abraunegg** [comment]
@eddy-geek 
Can you provide the entire contents of the `resume_upload` file that is in your `~/.config/onedrive` folder please - I want to validate what is currently in there

**Comment by abraunegg** [comment]
@eddy-geek 
I have been able to reproduce this by canceling out an upload & modifying the `resume_upload` file, however as I do not know what your file contains, unsure if this fix is complete.

Please can you test the following PR:
```
git clone https://github.com/abraunegg/onedrive.git
cd onedrive
git fetch origin pull/664/head:pr664
git checkout pr664
make clean; ./configure; make;
```

**Comment by abraunegg** [comment]
@eddy-geek
Any update as to testing the PR for this issue?

**Comment by abraunegg** [comment]
@eddy-geek
Any update as to testing the PR for this issue?

**Comment by eddy-geek** [comment]
Hello,

I don't have a resume_upload file (I think I deleted everything but refresh_token in an attempt to get it to work).

I just tested your PR and it seems to work!

It does crash later with "Operation not permitted" but that's another issue, linked to the ntfs mount I guess. It would be nice if it was a bit more resilient though? Sometimes there's just one file with permission issues.

Log:

```
Sep 19 09:16:15 ncelrnd0216 systemd[1701]: Stopped OneDrive Free Client.
Sep 19 09:16:23 ncelrnd0216 systemd[1701]: Started OneDrive Free Client.
Sep 19 09:16:25 ncelrnd0216 onedrive[3752]: Initializing the Synchronization Engine ...
Sep 19 09:16:25 ncelrnd0216 onedrive[3752]: Initializing monitor ...
Sep 19 09:16:25 ncelrnd0216 onedrive[3752]: OneDrive monitor interval (seconds): 45
Sep 19 09:18:37 ncelrnd0216 onedrive[3752]: Processing 203 changes
Sep 19 09:18:41 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:18:46 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:18:48 ncelrnd0216 onedrive[3752]: Processing 204 changes
Sep 19 09:18:49 ncelrnd0216 onedrive[3752]: Processing 207 changes
Sep 19 09:18:52 ncelrnd0216 onedrive[3752]: Processing 204 changes
Sep 19 09:18:54 ncelrnd0216 onedrive[3752]: Processing 204 changes
Sep 19 09:18:56 ncelrnd0216 onedrive[3752]: Processing 204 changes
Sep 19 09:18:58 ncelrnd0216 onedrive[3752]: Processing 204 changes
Sep 19 09:19:01 ncelrnd0216 onedrive[3752]: Processing 204 changes
Sep 19 09:19:04 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:07 ncelrnd0216 onedrive[3752]: Processing 206 changes
Sep 19 09:19:11 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:14 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:18 ncelrnd0216 onedrive[3752]: Processing 206 changes
Sep 19 09:19:19 ncelrnd0216 onedrive[3752]: Processing 206 changes
Sep 19 09:19:21 ncelrnd0216 onedrive[3752]: Processing 206 changes
Sep 19 09:19:24 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:26 ncelrnd0216 onedrive[3752]: Processing 206 changes
Sep 19 09:19:28 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:29 ncelrnd0216 onedrive[3752]: Processing 212 changes
Sep 19 09:19:31 ncelrnd0216 onedrive[3752]: Processing 204 changes
Sep 19 09:19:33 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:35 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:37 ncelrnd0216 onedrive[3752]: Processing 204 changes
Sep 19 09:19:39 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:41 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:43 ncelrnd0216 onedrive[3752]: Processing 206 changes
Sep 19 09:19:45 ncelrnd0216 onedrive[3752]: Processing 206 changes
Sep 19 09:19:49 ncelrnd0216 onedrive[3752]: Processing 206 changes
Sep 19 09:19:51 ncelrnd0216 onedrive[3752]: Processing 204 changes
Sep 19 09:19:53 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:19:55 ncelrnd0216 onedrive[3752]: Processing 215 changes
Sep 19 09:19:57 ncelrnd0216 onedrive[3752]: Processing 215 changes
Sep 19 09:20:03 ncelrnd0216 onedrive[3752]: Processing 212 changes
Sep 19 09:20:04 ncelrnd0216 onedrive[3752]: Processing 211 changes
Sep 19 09:20:05 ncelrnd0216 onedrive[3752]: Processing 207 changes
Sep 19 09:20:07 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:20:08 ncelrnd0216 onedrive[3752]: Downloading file Cours/Cours evolutifs/Tle.Industrial revolution/0.1.test 1 avec txt ... Retry sync count: 1: Cours/Cours evolutifs/Tle.Industrial revolution/0.1.test 1 avec txt: Operation not permitted
Sep 19 09:20:09 ncelrnd0216 onedrive[3752]: Processing 207 changes
Sep 19 09:20:10 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:20:14 ncelrnd0216 onedrive[3752]: Downloading file Cours/Cours evolutifs/Tle.Industrial revolution/0.1.test un (4 versions) ... Retry sync count: 2: Cours/Cours evolutifs/Tle.Industrial revolution/0.1.test un (4 versions): Operation not permitted
Sep 19 09:20:16 ncelrnd0216 onedrive[3752]: Processing 207 changes
Sep 19 09:20:17 ncelrnd0216 onedrive[3752]: Processing 205 changes
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: Downloading file Cours/Cours evolutifs/Tle.Industrial revolution/0.1.test voc plus texte.pdf ... Giving up on sync after three attempts: Cours/Cours evolutifs/Tle.Industrial revolution/0.1.test voc plus texte.pdf: Operation not permitted
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: std.file.FileException@/home/ed/dlang/dmd-2.087.0/linux/bin64/../../src/phobos/std/file.d(1550): Cours/Cours evolutifs/Tle.Industrial revolution/0.1.test voc plus texte.pdf: Operation not permitted
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: ----------------
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: /home/ed/dlang/dmd-2.087.0/linux/bin64/../../src/phobos/std/file.d:272 @trusted bool std.file.cenforce!(bool).cenforce(bool, scope const(char)[], scope const(char)*, immutable(char)[], ulong) [0x5e9c34]
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: /home/ed/dlang/dmd-2.087.0/linux/bin64/../../src/phobos/std/file.d:261 @safe void std.file.setTimes!(immutable(char)[]).setTimes(immutable(char)[], std.datetime.systime.SysTime, std.datetime.systime.SysTime) [0x605526]
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: src/sync.d:1460 void sync.SyncEngine.downloadFileItem(itemdb.Item, immutable(char)[]) [0x639278]
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: src/sync.d:1284 void sync.SyncEngine.applyNewItem(itemdb.Item, immutable(char)[]) [0x6384ba]
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: src/sync.d:1226 void sync.SyncEngine.applyDifference(std.json.JSONValue, immutable(char)[], bool) [0x6380be]
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: src/sync.d:923 void sync.SyncEngine.applyDifferences(immutable(char)[], const(char)[]) [0x63693e]
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: src/sync.d:419 void sync.SyncEngine.applyDifferences() [0x634c28]
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: src/main.d:836 void main.performSync(sync.SyncEngine, immutable(char)[], bool, bool, bool, long, bool) [0x622ba7]
Sep 19 09:20:18 ncelrnd0216 onedrive[3752]: src/main.d:709 _Dmain [0x6222a3]
Sep 19 09:20:18 ncelrnd0216 systemd[1701]: onedrive.service: Main process exited, code=exited, status=1/FAILURE
Sep 19 09:20:18 ncelrnd0216 systemd[1701]: onedrive.service: Failed with result 'exit-code'.
Sep 19 09:20:21 ncelrnd0216 systemd[1701]: onedrive.service: Service RestartSec=3s expired, scheduling restart.
Sep 19 09:20:21 ncelrnd0216 systemd[1701]: onedrive.service: Scheduled restart job, restart counter is at 1.
Sep 19 09:20:21 ncelrnd0216 systemd[1701]: Stopped OneDrive Free Client.
Sep 19 09:20:21 ncelrnd0216 systemd[1701]: Started OneDrive Free Client.
Sep 19 09:20:23 ncelrnd0216 onedrive[3992]: Initializing the Synchronization Engine ...
Sep 19 09:20:23 ncelrnd0216 onedrive[3992]: Initializing monitor ...
...
```

**Comment by eddy-geek** [comment]
I can open a separate issue for  the Permission Denied if you want.
The permissions are `.rwxrwxrwx root root` so I don't what is causing issues. Maybe trying to set metadata like Created/Modified times?.

**Comment by abraunegg** [comment]
@eddy-geek 
Unsure why it would be failing if the permissions are 'root' - what is the client running as - root? You mention that this is also a NTFS mount - so potentially some sort of 'mount & sync' deadlock outside of the client is preventing the time for the file to be set - almost like a race condition.

Will update the PR to handle this as well, as it should not crash in that sort of situation

**Comment by abraunegg** [comment]
@eddy-geek 
Please can you rebuild your client from the updated PR #664 to validate the issue of setting the file times being resolved

**Comment by eddy-geek** [comment]
Hello,

I can confirm the last fix prevents the crash, and allows uploads to resume normally.

The issue with the "corrupted" file downloads is recurrent so error messages are logged every cycle (but thankfully not notified ;)). This is good enough for now, thank you for the work and for the reactivity!

**Comment by abraunegg** [comment]
@eddy-geek 
If you can provide debug logging on the corrupt files, that would be greatly appreciated.

**Comment by abraunegg** [comment]
Marking as fixed & closing as PR #644 is merged into master
